lasagne/layers/recurrent.py

```diff
@@ -0,0 +1,1239 @@
+by using a few reshape operations; please refer to the example below.
+
+Examples
+--------
+The following example demonstrates how recurrent layers can be easily mixed
+with feed-forward layers using :class:`ReshapeLayer` and how to build a
+network with variable batch size and number of time steps.
+
+>>> from lasagne.layers import *
+>>> num_inputs, num_units, num_classes = 10, 12, 5
+>>> # By setting the first two dimensions as None, we are allowing them to vary
+>>> # They correspond to batch size and sequence length, so we will be able to
+>>> # feed in batches of varying size with sequences of varying length.
+>>> l_inp = InputLayer((None, None, num_inputs))
+>>> # We can retrieve symbolic references to the input variable's shape, which
+>>> # we will later use in reshape layers.
+>>> batchsize, seqlen, _ = l_inp.input_var.shape
+>>> l_lstm = LSTMLayer(l_inp, num_units=num_units)
+>>> # In order to connect a recurrent layer to a dense layer, we need to
+>>> # flatten the first two dimensions (our "sample dimensions"); this will
+>>> # cause each time step of each sequence to be processed independently
+>>> l_shp = ReshapeLayer(l_lstm, (-1, num_units))
+>>> l_dense = DenseLayer(l_shp, num_units=num_classes)
+>>> # To reshape back to our original shape, we can use the symbolic shape
+>>> # variables we retrieved above.
+>>> l_out = ReshapeLayer(l_dense, (batchsize, seqlen, num_classes))
+"""
+import numpy as np
+import theano
+import theano.tensor as T
+from .. import nonlinearities
+from .. import init
+from ..utils import unroll_scan
+
+from .base import Layer
+from .input import InputLayer
+from .dense import DenseLayer
+from . import helper
+
+__all__ = [
+    "CustomRecurrentLayer",
+    "RecurrentLayer",
+    "Gate",
+    "LSTMLayer",
+    "GRULayer"
+]
+
+
+class CustomRecurrentLayer(Layer):
+    """
+    lasagne.layers.recurrent.CustomRecurrentLayer(incoming, input_to_hidden,
+    hidden_to_hidden, nonlinearity=lasagne.nonlinearities.rectify,
+    hid_init=lasagne.init.Constant(0.), backwards=False,
+    learn_init=False, gradient_steps=-1, grad_clipping=False,
+    unroll_scan=False, precompute_input=True, **kwargs)
+
+    A layer which implements a recurrent connection.
+
+    This layer allows you to specify custom input-to-hidden and
+    hidden-to-hidden connections by instantiating layer instances and passing
+    them on initialization.  The output shape for the provided layers must be
+    the same.  If you are looking for a standard, densely-connected recurrent
+    layer, please see :class:`RecurrentLayer`.  The output is computed
+    by
+
+    .. math ::
+        h_t = \sigma(f_i(x_t) + f_h(h_{t-1}))
+
+    Parameters
+    ----------
+    incoming : a :class:`lasagne.layers.Layer` instance or a tuple
+        The layer feeding into this layer, or the expected input shape.
+    input_to_hidden : :class:`lasagne.layers.Layer`
+        Layer which connects input to the hidden state (:math:`f_i`).
```

Comment:
> Haha, sorry, should have been more clear: I meant that all instances of helper.get_output in your code (the code of the PR, not the code in your comment!) provide an input expression for the network, so they won't throw an error (unlike the code in your comment).

Right, the code in my comment was to illustrate the fact that the incoming argument of the input-to-hidden and hidden-to-hidden connections 

> can't be a layer created with a tuple as incoming

which is what I was intending to illustrate with the example.  I think we're not talking about the same thing here.

> Yes, a lasagne.Layer instance can refer to a network, but it's not clear whether you mean a lasagne.Layer instance here or the concept of a neural network layer, capitalized because it's the beginning of the sentence. Sorry, I'm not trying to be nitpicking, just pointing out a possible misunderstanding!

Sure, I will figure out a better way to word it.

> That might be philosophical, but I do think there are networks in lasagne! We're even using that term a lot in the documentation. There's no network class in lasagne, sure, but we can still distinguish networks and layers in the documentation. (A glossary might be helpful, though.)

Yes, some kind of glossary might be helpful - after all, in the new mnist example, the entire network is referred to as `layer`: https://github.com/Lasagne/Lasagne/blob/master/examples/mnist.py#L139